components/eventing-controller/cmd/eventing-controller/main.go

```diff
@@ -1,89 +1,77 @@
 
+	// Parse flags.
+	var enableDebugLogs bool
 	var metricsAddr string
 	var resyncPeriod time.Duration
-	var enableDebugLogs bool
-	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
-	flag.DurationVar(&resyncPeriod, "reconcile-period", time.Minute*10, "Period between triggering of reconciling calls.")
+	var maxReconnects int
+	var reconnectWait time.Duration
+
 	flag.BoolVar(&enableDebugLogs, "enable-debug-logs", false, "Enable debug logs.")
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.DurationVar(&resyncPeriod, "reconcile-period", time.Minute*10, "Period between triggering of reconciling calls (BEB).")
+	flag.IntVar(&maxReconnects, "max-reconnects", 10, "Maximum number of reconnect attempts (NATS).")
+	flag.DurationVar(&reconnectWait, "reconnect-wait", time.Second, "Wait time between reconnect attempts (NATS).")
 	flag.Parse()
 
-	cfg := env.GetConfig()
-	if len(cfg.Domain) == 0 {
-		setupLog.Error(fmt.Errorf("env var DOMAIN should be a non-empty value"), "unable to start manager")
-		os.Exit(1)
-	}
+	// Instantiate configured commander.
+	var commander Commander
+
+	backend := strings.ToUpper(os.Getenv(ENV_BACKEND))
 
-	// cluster config
-	k8sConfig := ctrl.GetConfigOrDie()
-
-	ctrl.SetLogger(zap.New(zap.UseDevMode(enableDebugLogs)))
-	mgr, err := ctrl.NewManager(k8sConfig, ctrl.Options{
-		Scheme:             scheme,
-		MetricsBindAddress: metricsAddr,
-		Port:               9443,
-		SyncPeriod:         &resyncPeriod,
-	})
-	if err != nil {
-		setupLog.Error(err, "unable to start manager")
+	switch backend {
+	case ENV_BACKEND_VALUE_BEB:
+		commander = beb.NewCommander(enableDebugLogs, metricsAddr, resyncPeriod)
+	case ENV_BACKEND_VALUE_NATS:
+		commander = nats.NewCommander(enableDebugLogs, metricsAddr, maxReconnects, reconnectWait)
+	default:
+		logger.Error(fmt.Errorf("specified invalid eventing controller backend: '%s'", backend), "unable to start manager")
 		os.Exit(1)
 	}
 
-	// setup application lister
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	dynamicClient := dynamic.NewForConfigOrDie(k8sConfig)
-	applicationLister := application.NewLister(ctx, dynamicClient)
-
-	if err := subscription.NewReconciler(
-		mgr.GetClient(),
-		applicationLister,
-		mgr.GetCache(),
-		ctrl.Log.WithName("reconciler").WithName("Subscription"),
-		mgr.GetEventRecorderFor("eventing-controller"),
-		cfg,
-	).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to setup the Subscription controller")
-		cancel()
+	// Init and start the commander.
+	if err := commander.Init(); err != nil {
+		logger.Error(err, "unable to start manager")
 		os.Exit(1)
 	}
 
-	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "unable to start manager")
-		cancel()
+	logger.Info(fmt.Sprintf("starting %v subscription controller and manager", commander))
```

Comment:
The `%v` will show the default format (which is the internal view for the impl struct). Can we improve this please?

Reply:
Ouch, has been intended to show the chosen backend.